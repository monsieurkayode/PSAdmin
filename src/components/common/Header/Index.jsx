import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { string, shape, arrayOf } from 'prop-types';

import brandLogo from '../../../images/pluralsight-logo.png';
import '../../../styles/Header.scss';

/**
 *
 * @function Header
 *
 * @param {object} location
 *
 * @returns {JSX}
 */
export const Header = ({ location: { pathname }, courses, authors }) => {
  /**
   * @description sets the active class
   * on the navbar item of the current page
   *
   * @function activeLink
   *
   * @param {string} path
   */
  const activeLink = (path) => {
    if (path.length === 1) {
      return pathname === '/';
    }

    return pathname.includes(path);
  };

  return (
    <nav className="navbar navbar-default">
      <div className="container">
        <Link to="/" className="navbar-brand">
          <img src={brandLogo} alt="brand" />
        </Link>
        <ul className="nav navbar-nav pull-right">
          <li className={`link ${activeLink('/') && 'active'}`}>
            <Link to="/">Home</Link>
          </li>
          <li className={`link ${activeLink('author') && 'active'}`}>
            <Link to="/authors">
              Authors <span className="count">{authors.length}</span>
            </Link>
          </li>
          <li className={`link ${activeLink('course') && 'active'}`}>
            <Link to="/courses">
              Courses <span className="count">{courses.length}</span>
            </Link>
          </li>
          <li className={`link ${activeLink('about') && 'active'}`}>
            <Link to="/about">About</Link>
          </li>
        </ul>
      </div>
    </nav>
  );
};

/**
 * @function mapStateToProps
 *
 * @param {object} state
 *
 * @returns {object}
 */
const mapStateToProps = ({
  allCourses: { courses },
  allAuthors: { authors }
}) => ({ courses, authors });

Header.propTypes = {
  location: shape({
    pathname: string.isRequired
  }).isRequired,
  courses: arrayOf(shape({})).isRequired,
  authors: arrayOf(shape({})).isRequired,
};

export default withRouter(connect(mapStateToProps)(Header));
