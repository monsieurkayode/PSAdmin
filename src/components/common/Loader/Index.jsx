import React from 'react';
import PreloaderIcon from 'react-preloader-icon';
import Oval from 'react-preloader-icon/loaders/Oval';

import '../../../styles/Loader.scss';

/**
 *
 * Loader
 * @function Loader
 *
 * @return {JSX} JSX
 */
const Loader = () => (
  <div className="loader">
    <PreloaderIcon
      loader={Oval}
      size={40}
      strokeWidth={12}
      strokeColor="#337AB7"
      duration={800}
    />
  </div>
);

export default Loader;
