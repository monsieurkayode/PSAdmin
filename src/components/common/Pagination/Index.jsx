import React from 'react';
import ReactPaginate from 'react-paginate';
import { number } from 'prop-types';

const Pagination = ({ pageCount, page, handlePageChange }) => ( // eslint-disable-line
  <div style={{ textAlign: 'center' }}>
    <ReactPaginate
      breakLabel={<span>...</span>}
      breakClassName="break-me"
      pageCount={pageCount}
      initialPage={page - 1}
      marginPagesDisplayed={2}
      pageRangeDisplayed={3}
      onPageChange={handlePageChange}
      disableInitialCallback
      containerClassName="pagination"
      subContainerClassName="pages pagination"
      activeClassName="active"
    />
  </div>
);

Pagination.propTypes = {
  pageCount: number.isRequired,
  page: number.isRequired
};

export default Pagination;
