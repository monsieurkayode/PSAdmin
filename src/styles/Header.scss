.link {
  transition: background 0.3s linear;
}

.link:hover,
.active {
  color: #000;
  background: #eee;
}

.count {
  display: inline-block;
  text-align: center;
  width: 20px;
  color: #fff;
  background: #0a73b2;
  border-radius: 50%;
}
