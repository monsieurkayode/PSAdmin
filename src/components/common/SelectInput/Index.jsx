import React from 'react';
import {
  string, func, arrayOf, shape, bool
} from 'prop-types';
import Loader from '../Loader/Index';

/**
 * @function SelectInput
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const SelectInput = ({
  name,
  label,
  defaultOption,
  onChange,
  value,
  error,
  options,
  loading,
  onFocus
}) => (
  <div className="form-group">
    <label htmlFor={name}>{label}</label>
    <div className="field custom-field">
      <select
        name={name}
        className="form-control"
        onChange={onChange}
        onFocus={onFocus}
        value={value}
      >
        <option value="">{defaultOption}</option>
        {options.map(option => (
          <option key={option.value} value={option.value}>{option.text}</option>
        ))}
      </select>
      {loading && <Loader size={20} />}
    </div>
    {error && (
      <div
        style={{ maxWidth: '93%', padding: '5px' }}
        className="alert alert-danger"
      >
        {error}
      </div>
    )}
  </div>
);

SelectInput.defaultProps = {
  value: '',
  error: ''
};

SelectInput.propTypes = {
  name: string.isRequired,
  label: string.isRequired,
  defaultOption: string.isRequired,
  onChange: func.isRequired,
  options: arrayOf(shape({})).isRequired,
  value: string,
  error: string,
  loading: bool.isRequired,
  onFocus: func.isRequired
};

export default SelectInput;
