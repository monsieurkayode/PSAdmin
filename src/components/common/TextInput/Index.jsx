import React from 'react';
import { string, func, bool } from 'prop-types';
import Loader from '../Loader/Index';

/**
 * @function TextInput
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const TextInput = ({
  name,
  label,
  placeholder,
  onChange,
  onFocus,
  value,
  error,
  loading
}) => {
  let wrapperClass = 'form-group';
  if (error && error.length > 0) {
    wrapperClass = `${wrapperClass} has-error`;
  }

  return (
    <div className={wrapperClass}>
      <label htmlFor={name}>{label}</label>
      <div className="field custom-field">
        <input
          type="text"
          name={name}
          className="form-control"
          placeholder={placeholder}
          onChange={onChange}
          onFocus={onFocus}
          value={value}
        />
        {loading && <Loader size={20} />}
      </div>
      {error && (
        <div
          style={{ maxWidth: '93%', padding: '5px' }}
          className="alert alert-danger"
        >
          {error}
        </div>
      )}
    </div>
  );
};

TextInput.defaultProps = {
  placeholder: '',
  value: '',
  error: ''
};

TextInput.propTypes = {
  name: string.isRequired,
  label: string.isRequired,
  onChange: func.isRequired,
  onFocus: func.isRequired,
  placeholder: string,
  value: string,
  error: string,
  loading: bool.isRequired
};

export default TextInput;
