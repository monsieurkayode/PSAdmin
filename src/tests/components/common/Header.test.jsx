import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../../components/common/Header/Index';

const setup = () => {
  const props = {
    location: {
      pathname: '/'
    },
    courses: [],
    authors: []
  };

  return shallow(<Header {...props} />);
};

describe('Header Component', () => {
  const wrapper = setup();

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has one ul element with four list items', () => {
    expect(wrapper.find('ul').exists()).toBe(true);
    expect(wrapper.find('ul').children().length).toBe(4);
  });

  it('adds class of active to nav-item that matches page route', () => {
    expect(wrapper.find('.active').length).toEqual(1);
  });
});
