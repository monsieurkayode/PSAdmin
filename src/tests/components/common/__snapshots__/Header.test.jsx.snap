// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders correctly 1`] = `
<nav
  className="navbar navbar-default"
>
  <div
    className="container"
  >
    <Link
      className="navbar-brand"
      replace={false}
      to="/"
    >
      <img
        alt="brand"
        src="test-file-mock"
      />
    </Link>
    <ul
      className="nav navbar-nav pull-right"
    >
      <li
        className="link active"
      >
        <Link
          replace={false}
          to="/"
        >
          Home
        </Link>
      </li>
      <li
        className="link false"
      >
        <Link
          replace={false}
          to="/authors"
        >
          Authors 
          <span
            className="count"
          >
            0
          </span>
        </Link>
      </li>
      <li
        className="link false"
      >
        <Link
          replace={false}
          to="/courses"
        >
          Courses 
          <span
            className="count"
          >
            0
          </span>
        </Link>
      </li>
      <li
        className="link false"
      >
        <Link
          replace={false}
          to="/about"
        >
          About
        </Link>
      </li>
    </ul>
  </div>
</nav>
`;
